File: ThreeDSecure/src/main/java/com/braintreepayments/api/threedsecure/ThreeDSecureNonce.java
package com.braintreepayments.api;


import android.os.Parcel;

import androidx.annotation.NonNull;
import androidx.annotation.RestrictTo;

import com.braintreepayments.api.card.AuthenticationInsight;
import com.braintreepayments.api.card.BinData;
import com.braintreepayments.api.card.CardNonce;

import org.json.JSONException;
import org.json.JSONObject;

/**
 * {@link PaymentMethodNonce} representing a {@link CardNonce} that includes {@link ThreeDSecureInfo}
 */
public class ThreeDSecureNonce extends CardNonce {
    private static final String THREE_D_SECURE_INFO_KEY = "threeDSecureInfo";
    private final ThreeDSecureInfo threeDSecureInfo;

    private ThreeDSecureNonce(CardNonce cardNonce, ThreeDSecureInfo threeDSecureInfo) {
        this(threeDSecureInfo, cardNonce.getCardType(), cardNonce.getLastTwo(),
                cardNonce.getLastFour(), cardNonce.getBin(), cardNonce.getBinData(),
                cardNonce.getAuthenticationInsight(), cardNonce.getExpirationMonth(),
                cardNonce.getExpirationYear(), cardNonce.getCardholderName(), cardNonce.getString(),
                cardNonce.isDefault());
    }

    private ThreeDSecureNonce(ThreeDSecureInfo threeDSecureInfo, String cardType, String lastTwo,
                              String lastFour, String bin, BinData binData,
                              AuthenticationInsight authenticationInsight, String expirationMonth,
                              String expirationYear, String cardholderName, String nonce,
                              boolean isDefault) {
        super(cardType, lastTwo, lastFour, bin, binData, authenticationInsight, expirationMonth,
                expirationYear, cardholderName, nonce, isDefault);
        this.threeDSecureInfo = threeDSecureInfo;
    }

    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    @NonNull
    public static ThreeDSecureNonce fromJSON(JSONObject inputJson) throws JSONException {
        CardNonce cardNonce = CardNonce.fromJSON(inputJson);
        ThreeDSecureInfo threeDSecureInfo;
        if (inputJson.has(DATA_KEY)) { // graphQL
           threeDSecureInfo = ThreeDSecureInfo.fromJson(null);
        } else if (inputJson.has(API_RESOURCE_KEY)) { // REST
            JSONObject json = inputJson.getJSONArray(API_RESOURCE_KEY).getJSONObject(0);
            threeDSecureInfo =
                    ThreeDSecureInfo.fromJson(json.optJSONObject(THREE_D_SECURE_INFO_KEY));
        } else { // plain JSON
            threeDSecureInfo =
                    ThreeDSecureInfo.fromJson(inputJson.optJSONObject(THREE_D_SECURE_INFO_KEY));
        }
        return new ThreeDSecureNonce(cardNonce, threeDSecureInfo);
    }

    /**
     * @return The 3D Secure info for the current {@link ThreeDSecureNonce} or {@code null}
     */
    @NonNull
    public ThreeDSecureInfo getThreeDSecureInfo() {
        return threeDSecureInfo;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeParcelable(threeDSecureInfo, flags);
    }

    protected ThreeDSecureNonce(Parcel in) {
        super(in);
        threeDSecureInfo = in.readParcelable(ThreeDSecureInfo.class.getClassLoader());
    }

    public static final Creator<ThreeDSecureNonce> CREATOR = new Creator<>() {
        public ThreeDSecureNonce createFromParcel(Parcel source) {
            return new ThreeDSecureNonce(source);
        }

        public ThreeDSecureNonce[] newArray(int size) {
            return new ThreeDSecureNonce[size];
        }
    };
}

File: ThreeDSecure/src/main/java/com/braintreepayments/api/threedsecure/ThreeDSecureNonce.kt
package com.braintreepayments.api.threedsecure

import androidx.annotation.RestrictTo
import com.braintreepayments.api.card.AuthenticationInsight
import com.braintreepayments.api.card.BinData
import com.braintreepayments.api.card.CardNonce
import com.braintreepayments.api.sharedutils.Json
import kotlinx.parcelize.Parcelize
import org.json.JSONException
import org.json.JSONObject

/**
 * [PaymentMethodNonce] representing a [cardNonce] that includes [ThreeDSecureInfo]
 *
 * @property threeDSecureInfo The 3D Secure info for the current [ThreeDSecureNonce] or `null`
 */
@Parcelize
data class ThreeDSecureNonce internal constructor(
    override val cardType: String,
    override val lastTwo: String,
    override val lastFour: String,
    override val bin: String,
    override val binData: BinData,
    override val authenticationInsight: AuthenticationInsight?,
    override val expirationMonth: String,
    override val expirationYear: String,
    override val cardholderName: String,
    override val string: String,
    override val isDefault: Boolean,
    val threeDSecureInfo: ThreeDSecureInfo
) : CardNonce(
    cardType = cardType,
    lastTwo = lastTwo,
    lastFour = lastFour,
    bin = bin,
    binData = binData,
    authenticationInsight = authenticationInsight,
    expirationMonth = expirationMonth,
    expirationYear = expirationYear,
    cardholderName = cardholderName,
    string = string,
    isDefault = isDefault
) {

    companion object {
        private const val THREE_D_SECURE_INFO_KEY = "threeDSecureInfo"

        @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
        @Throws(JSONException::class)
        @JvmStatic
        fun fromJSON(json: JSONObject): ThreeDSecureNonce {
            val cardNonce = CardNonce.fromJSON(json)
            val threeDSecureInfo: ThreeDSecureInfo
            if (json.has(DATA_KEY)) { // graphQL
                threeDSecureInfo = ThreeDSecureInfo.fromJson(null)
            } else if (json.has(API_RESOURCE_KEY)) { // REST
                val json = json.getJSONArray(API_RESOURCE_KEY).getJSONObject(0)
                threeDSecureInfo =
                    ThreeDSecureInfo.fromJson(json.optJSONObject(THREE_D_SECURE_INFO_KEY))
            } else { // plain JSON
                threeDSecureInfo = ThreeDSecureInfo.fromJson(json.optJSONObject(THREE_D_SECURE_INFO_KEY))
            }
            return ThreeDSecureNonce(
                cardType = cardNonce.cardType,
                lastTwo = cardNonce.lastTwo,
                lastFour = cardNonce.lastFour,
                bin = cardNonce.bin,
                binData = cardNonce.binData,
                authenticationInsight = cardNonce.authenticationInsight,
                expirationMonth = cardNonce.expirationMonth,
                expirationYear = cardNonce.expirationYear,
                cardholderName = cardNonce.cardholderName,
                string = cardNonce.string,
                isDefault = cardNonce.isDefault,
                threeDSecureInfo = threeDSecureInfo
            )
        }
    }
}
